src/query.rs

```diff
@@ -422,30 +422,42 @@ impl Query {
         Ok(blockhash)
     }
 
-    // Internal API for transaction retrieval (uses bitcoind)
-    fn _load_txn(&self, tx_hash: &Sha256dHash, block_height: u32) -> Result<Transaction> {
-        let blockhash = self.lookup_confirmed_blockhash(tx_hash, Some(block_height))?;
-        self.app.daemon().gettransaction(tx_hash, blockhash)
-    }
-
-    // Get transaction from txstore or the in-memory mempool Tracker
-    pub fn tx_get(&self, txid: &Sha256dHash) -> Option<Transaction> {
-        rawtxrow_by_txid(self.app.read_store(), txid)
-            .map(|row| deserialize(&row.rawtx).expect("cannot parse tx from txstore"))
-            .or_else(|| self.tracker.read().unwrap().get_txn(&txid))
+    // Load transaction by txid
+    pub fn load_txn(&self, txid: &Sha256dHash, blockhash: Option<&Sha256dHash>) -> Result<Transaction> {
+        if self.app.config().txstore_enabled {
+            // fetch from our txstore or mempool tracker
+            rawtxrow_by_txid(self.app.read_store(), txid)
+                .map(|row| deserialize(&row.rawtx).expect("cannot parse tx from txstore"))
+                .or_else(|| self.tracker.read().unwrap().get_txn(&txid))
+                .chain_err(|| format!("cannot find tx {}", txid))
+        } else {
+            // fetch from bitcoind
+            let blockhash = match blockhash {
+                Some(x) => Some(*x),
```

Comment:
Please take a look at https://github.com/romanz/electrs/commit/649f266ac480c1f51b5781ba3c1dbfe806cabc89 (IIUC it should not involve copying the Sha256dHash).

Reply:
Will implemented and force-push over the previous commit.

Question: any idea why the compiler requires an explicit type annotation for `blockhash_from_index`?

Reply:
Implemented in 81954752041d012de542cf7b3a754f8ac0033e3c.

Reply:
Thanks!

> Question: any idea why the compiler requires an explicit type annotation for `blockhash_from_index`?

Actually, I think that it's fine to omit the type (checked now with stable Rust 1.30.0), so feel free to remove it.
I've added it initially to be more explicit about the actual object and its reference, but I'm not sure the verbosity is worth it.